### PR TITLE
Sm/acccesstoken works as username

### PR DESCRIPTION
**Pull request by mshanemc**
@W-7885939@

By fixing this accessToken issue in Core, it might also solve some other things like
- https://github.com/forcedotcom/cli/issues/290

### CODE REVIEW

File: src/authInfo.ts [hunk 1]
```diff
@@ -758,8 +759,7 @@ export class AuthInfo extends AsyncCreatable<AuthInfo.Options> {
     this.fields.username = this.options.username || getString(options, 'username') || undefined;
 
     // If the username is an access token, use that for auth and don't persist
-    const accessTokenMatch = isString(this.fields.username) && this.fields.username.match(/^(00D\w{12,15})![.\w]*$/);
-    if (accessTokenMatch) {
+    if (isString(this.fields.username) && sfdc.matchesAccessToken(this.fields.username)) {
```
Comment (mshanemc) [inline src/authInfo.ts:762]: used in multiple places, so moved to the sfdc utils

File: src/user.ts [hunk 2]
```diff
@@ -75,6 +75,14 @@ async function retrieveUserFields(logger: Logger, username: string): Promise<Use
     authInfo: await AuthInfo.create({ username }),
   });
 
+  if (sfdc.matchesAccessToken(username)) {
```
Comment (WillieRuemmele) [inline src/user.ts:78]: should resolving the username like this be used in creating the Connection above?

File: src/user.ts [hunk 3]
```diff
@@ -75,6 +75,14 @@ async function retrieveUserFields(logger: Logger, username: string): Promise<Use
...
+    logger.debug('received an accessToken for the username.  Converting...');
+    username = (await connection.identity()).username;
+    logger.debug(`accessToken converted to ${username}`);
+  } else {
+    logger.debug('not a accessToken');
+  }
+
   const fromFields = Object.keys(REQUIRED_FIELDS).map(upperFirst);
```
Comment (peternhale) [inline src/user.ts:86]: Not part of this change, but I do not see the sense in the transformation of required fields to name case.

@shetzel, is there a reason for this?

File: src/authInfo.ts [hunk 4]
```diff
@@ -758,8 +759,7 @@ export class AuthInfo extends AsyncCreatable<AuthInfo.Options> {
     this.fields.username = this.options.username || getString(options, 'username') || undefined;
```
Comment (peternhale) [inline src/authInfo.ts:759]: Is last `|| undefined` necessary, given getString either returns the property or undefined?

File: src/util/sfdc.ts [hunk 5]
```diff
@@ -111,4 +111,13 @@ export const sfdc = {
...
     return key;
   },
+
+  /**
+   * Tests whether a given string is an access token
+   *
+   * @param value
+   */
+  matchesAccessToken: (value: string): boolean => {
+    return value.match(/^(00D\w{12,15})![.\w]*$/) !== null;
```
Comment (peternhale) [inline src/util/sfdc.ts:121]: Could use `/^(00D\w{12,15})![.\w]*$/.test(value);`

**Review by peternhale — APPROVED** [review]
With some suggestions.

File: src/util/sfdc.ts (hunk 5 above)
Comment (mshanemc) [reply src/util/sfdc.ts:121]: agree!

File: src/user.ts (hunk 2 above)
Comment (mshanemc) [reply src/user.ts:78]: I don't understand Auth well enough to know, so I'm limiting the change to fixing the bug.

It definitely works OK to get the Connection via accessToken.

File: src/authInfo.ts (hunk 4 above)
Comment (shetzel) [reply src/authInfo.ts:759]: I agree that the `|| undefined` suffix looks unnecessary based on how `getString` works

File: src/user.ts (hunk 2 above)
Comment (shetzel) [reply src/user.ts:78]: No we must use the access token for the connection if it's passed in.  If we used the username it would fail because there's not associated auth file.

File: src/user.ts (hunk 3 above)
Comment (shetzel) [reply src/user.ts:86]: There was a reason for it at the time.  I'm guessing it was something to do with the User API.

**Review by shetzel — APPROVED** [review]

File: src/authInfo.ts (hunk 4 above)
Comment (mshanemc) [reply src/authInfo.ts:759]: TS complains:
_Type 'Optional<string | null>' is not assignable to type 'string | undefined'.
  Type 'null' is not assignable to type 'string | undefined'._

going to leave it for now

Comment (shetzel) [reply src/authInfo.ts:759]: Ahh right.  `getString` returns `string | null` not `string | undefined`.  null !== undefined
